File: namui/src/render/Clip.ts
import { ClipOp, Path } from "canvaskit-wasm";
import { RenderingTree } from "../type";
import { ClipCommand } from "../types/SpecialRenderingCommand";

export function Clip(
  {
    path,
    clipOp,
  }: {
    path: Path;
    clipOp: ClipOp;
  },
  renderingTree: RenderingTree,
): ClipCommand {
  return {
    type: "clip",
    path,
    clipOp,
    renderingTree,
  };
}

File: namui/src/namui/render/clip.rs
use super::{RenderingTree, SpecialRenderingNode};
use crate::namui::{ClipOp, Path};
use serde::Serialize;

#[derive(Serialize, Clone)]
pub struct Clip {
    pub(crate) path: Path,
    pub(crate) clip_op: ClipOp,
    pub(crate) rendering_tree: Vec<RenderingTree>,
}

pub fn clip(path: Path, clip_op: ClipOp, rendering_tree: RenderingTree) -> RenderingTree {
    RenderingTree::Special(SpecialRenderingNode::Clip(Clip {
        path,
        clip_op,
        rendering_tree: vec![rendering_tree],
    }))
}
